### PR TITLE
time-descending result of select queries

### PR DESCRIPTION
**Pull request by navis**
#2014 for select query

### CODE REVIEW

File: processing/src/main/java/io/druid/query/select/PagingOffset.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  {
+    counter++;
+  }
+
+  public int current()
+  {
+    return offset + (descending ? -counter : counter);
+  }
+
+  static int toDelta(int offset, boolean descending)
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/select/PagingOffset.java]: can we have some different names between these offsets and the one defined in the class?

Comment (navis) [inline processing/src/main/java/io/druid/query/select/PagingOffset.java]: renamed offset in the class to startOffset

**Comment by fjy** [comment]
:+1: 

@navis please add documentation for select query that it is reversible

**Comment by navis** [comment]
updated document

File: processing/src/main/java/io/druid/query/Druids.java [hunk 2]
```diff
@@ -1134,6 +1136,12 @@ public SelectQueryBuilder intervals(List<Interval> l)
       return this;
     }
 
+    public SelectQueryBuilder descending(boolean descending)
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/Druids.java]: `setDescending`

**Comment by fjy** [comment]
@himanshug can you take a look?

File: processing/src/main/java/io/druid/query/select/SelectQuery.java [hunk 3]
```diff
@@ -65,6 +66,17 @@ public SelectQuery(
     this.pagingSpec = pagingSpec;
 
     Preconditions.checkNotNull(pagingSpec, "must specify a pagingSpec");
+    Preconditions.checkArgument(checkPagingSpec(pagingSpec, descending), "invalid pagingSpec");
+  }
+
+  private boolean checkPagingSpec(PagingSpec pagingSpec, boolean descending)
+  {
+    for (Integer value : pagingSpec.getPagingIdentifiers().values()) {
+      if (descending ^ value < 0) {
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/select/SelectQuery.java]: This is a part of the java spec I'm not as familiar with. Is this doing `((descending ? 1 : 0) ^ value) < 0` or `descending ^ (value < 0)`?

File: processing/src/main/java/io/druid/query/Druids.java (hunk 2 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/Druids.java]: I see other methods in this class violate "normal" bean naming conventions. Oh well...

File: processing/src/main/java/io/druid/query/select/SelectQuery.java (hunk 3 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/select/SelectQuery.java]: IntelliJ says it's `descending ^ (value < 0)`. I'll add clarifying parentheses.

File: processing/src/main/java/io/druid/query/Druids.java (hunk 2 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/Druids.java]: @drcrallen any other comments? I think this is pretty close to ready

Comment (fjy) [inline processing/src/main/java/io/druid/query/Druids.java]: Also I think in Druids and in other places, we did not follow the 'set' convention in most of the builder classes.

File: processing/src/main/java/io/druid/query/select/SelectQueryEngine.java [hunk 4]
```diff
@@ -102,18 +102,11 @@
...
-              startOffset = (offset == null) ? 0 : offset;
-            }
+            final PagingOffset offset = query.getPagingOffset(segment.getIdentifier());
 
-            cursor.advanceTo(startOffset);
+            cursor.advanceTo(offset.startDelta());
 
-            int offset = 0;
-            while (!cursor.isDone() && offset < query.getPagingSpec().getThreshold()) {
+            for (; !cursor.isDone() && offset.hasNext(); cursor.advance(), offset.next()) {
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/select/SelectQueryEngine.java]: this is so much tidier now

File: docs/content/development/select-query.md [hunk 5]
```diff
@@ -25,6 +26,7 @@ There are several main parts to a select query:
 |queryType|This String should always be "select"; this is the first thing Druid looks at to figure out how to interpret the query|yes|
 |dataSource|A String or Object defining the data source to query, very similar to a table in a relational database. See [DataSource](../querying/datasource.html) for more information.|yes|
 |intervals|A JSON Object representing ISO-8601 Intervals. This defines the time ranges to run the query over.|yes|
+|descending|Whether to make descending ordered result. Default is `false`(ascending).|no|
```
Comment (drcrallen) [inline docs/content/development/select-query.md]: If I'm reading this right, the paging spec needs negative identifiers if this is set to true. Can that be documented?

File: processing/src/main/java/io/druid/query/select/PagingOffset.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ * under the License.
+ */
+
+package io.druid.query.select;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ */
+public abstract class PagingOffset
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/select/PagingOffset.java]: Can you add tests for this class (without relying on SelectQueryRunnerTest)?

**Comment by drcrallen** [comment]
Functionality looks good. Have some outstanding questions about docs and possibly some more unit tests.

**Comment by drcrallen** [comment]
Also please squash soon, this is very close to ready

File: docs/content/development/select-query.md (hunk 5 above)
Comment (navis) [inline docs/content/development/select-query.md]: sure

File: processing/src/main/java/io/druid/query/select/PagingOffset.java (hunk 6 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/select/PagingOffset.java]: ok

File: docs/content/development/select-query.md (hunk 5 above)
Comment (nishantmonu51) [inline docs/content/development/select-query.md]: Also, while paging the client increments the offset by 1 and queries again for next set of results, 
is it true for desending queries also or the client needs to decrement the offset by 1 ?  
This needs to be made clear in the docs.

**Comment by fjy** [comment]
@navis can you squash commits?

**Comment by navis** [comment]
squashed

**Comment by fjy** [comment]
@drcrallen any more comments?

**Comment by drcrallen** [comment]
is not squashed :) did @navis forget to push?

**Comment by fjy** [comment]
@navis ?

**Comment by navis** [comment]
Ah, sorry. I've confused with other issue.

**Comment by navis** [comment]
I found I've lost one commit, yesterday. I can't say it's related to shutdown of github.

**Comment by drcrallen** [comment]
:+1:

**Pull request merged**